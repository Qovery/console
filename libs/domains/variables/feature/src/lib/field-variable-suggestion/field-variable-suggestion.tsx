import { useRef } from 'react'
import { Button, Icon, InputTextSmall, type InputTextSmallProps } from '@qovery/shared/ui'
import { twMerge } from '@qovery/shared/util-js'
import DropdownVariable from '../dropdown-variable/dropdown-variable'

export interface FieldVariableSuggestionProps {
  environmentId: string
  onChange: (value: string) => void
  inputProps: InputTextSmallProps
  value?: string
  className?: string
}

export function FieldVariableSuggestion({
  environmentId,
  onChange,
  className,
  inputProps,
  value = '',
}: FieldVariableSuggestionProps) {
  const inputRef = useRef<HTMLInputElement | null>(null)

  const handleInsertVariable = (variableKey: string) => {
    const input = inputRef.current
    if (!input) return

    const startPos = input.selectionStart ?? 0
    const endPos = input.selectionEnd ?? 0

    const newValue = value.substring(0, startPos) + `{{${variableKey}}}` + value.substring(endPos)

    onChange(newValue)
  }

  // XXX: https://github.com/radix-ui/primitives/issues/1342
  // We are waiting for radix combobox primitives
  // So we are using DropdownMenu.Root in combination of Popover.Root
  // to get the flexibility of Popover.Root but keeping the accessiblity of
  // DropdownMenu.Root for entries.
  // So both open state should be sync
  return (
    <div className={twMerge('flex w-[calc(100%+36px)]', className)}>
      <InputTextSmall ref={inputRef} {...inputProps} inputClassName="pr-11" />
      <DropdownVariable environmentId={environmentId} onChange={handleInsertVariable}>
        <Button
          size="md"
          type="button"
          color="neutral"
          variant="surface"
          className="relative right-[37px] top-[1px] h-[34px] w-9 justify-center rounded-l-none border-b-0 border-r-0 border-t-0 hover:!border-neutral-250"
        >
          <Icon className="text-sm" iconName="wand-magic-sparkles" />
        </Button>
      </DropdownVariable>
    </div>
  )
}

export default FieldVariableSuggestion
