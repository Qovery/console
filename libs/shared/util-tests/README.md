# shared-util-tests

This library was generated with [Nx](https://nx.dev).

## Running unit tests

Run `nx test shared-util-tests` to execute the unit tests via [Jest](https://jestjs.io).
