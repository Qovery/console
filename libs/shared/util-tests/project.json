{
  "name": "shared-util-tests",
  "$schema": "../../../node_modules/nx/schemas/project-schema.json",
  "sourceRoot": "libs/shared/util-tests/src",
  "projectType": "library",
  "tags": ["scope:shared", "type:util"],
  "implicitDependencies": ["!shared-ui"],
  "targets": {
    "test": {
      "options": {
        "passWithNoTests": true
      },
      "configurations": {
        "ci": {
          "ci": true,
          "coverage": true
        }
      }
    }
  }
}
