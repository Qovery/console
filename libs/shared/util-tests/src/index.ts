export * from './lib/shared-util-tests'
