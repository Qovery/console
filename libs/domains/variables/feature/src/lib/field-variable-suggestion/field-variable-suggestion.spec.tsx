import { renderWithProviders, screen } from '@qovery/shared/util-tests'
import { FieldVariableSuggestion, type FieldVariableSuggestionProps } from './field-variable-suggestion'

const props: FieldVariableSuggestionProps = {
  environmentId: '000',
  onChange: jest.fn(),
}

jest.mock('../hooks/use-variables/use-variables', () => ({
  useVariables: () => ({
    data: [
      {
        id: 'c0277184-1fe0-4f17-b09a-58eee2c05701',
        created_at: '2023-10-25T09:13:34.723567Z',
        updated_at: '2023-10-25T09:13:34.723568Z',
        key: 'QOVERY_CONTAINER_Z04308DE2_HOST_INTERNAL',
        value: 'app-z04308de2-back-end',
        mount_path: null,
        scope: 'BUILT_IN',
        overridden_variable: null,
        aliased_variable: null,
        variable_type: 'BUILT_IN',
        service_id: '04308de2-af27-405f-9e95-570fa94ed577',
        service_name: 'back-end-A',
        service_type: 'CONTAINER',
        owned_by: 'QOVERY',
        description: 'test',
        is_secret: false,
      },
    ],
  }),
}))

describe('FieldVariableSuggestion', () => {
  it('should render successfully', () => {
    const { baseElement } = renderWithProviders(<FieldVariableSuggestion {...props} />)
    expect(baseElement).toBeTruthy()
  })

  it('should render dropdown content when clicked', async () => {
    const { userEvent } = renderWithProviders(<FieldVariableSuggestion {...props} />)

    const button = screen.getByRole('button')
    await userEvent.click(button)

    const dropdownContent = await screen.findByRole('menu')

    expect(dropdownContent).toMatchSnapshot()
  })
})
