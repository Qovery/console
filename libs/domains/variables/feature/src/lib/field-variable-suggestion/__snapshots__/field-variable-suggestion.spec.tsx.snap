// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FieldVariableSuggestion should render dropdown content when clicked 1`] = `
<div
  aria-labelledby="radix-:rf:"
  aria-orientation="vertical"
  class="rounded border border-neutral-200 bg-white shadow-xl outline-none data-[state=open]:data-[side=bottom]:animate-slidein-up-sm-faded data-[state=open]:data-[side=left]:animate-slidein-right-sm-faded data-[state=open]:data-[side=right]:animate-slidein-left-sm-faded data-[state=open]:data-[side=top]:animate-slidein-down-sm-faded flex max-h-60 w-[256px] flex-col p-2"
  data-align="center"
  data-orientation="vertical"
  data-radix-menu-content=""
  data-side="bottom"
  data-state="open"
  dir="ltr"
  id="radix-:rg:"
  role="menu"
  style="outline: none; --radix-dropdown-menu-content-transform-origin: var(--radix-popper-transform-origin); --radix-dropdown-menu-content-available-width: var(--radix-popper-available-width); --radix-dropdown-menu-content-available-height: var(--radix-popper-available-height); --radix-dropdown-menu-trigger-width: var(--radix-popper-anchor-width); --radix-dropdown-menu-trigger-height: var(--radix-popper-anchor-height); --radix-popover-content-transform-origin: var(--radix-popper-transform-origin); --radix-popover-content-available-width: var(--radix-popper-available-width); --radix-popover-content-available-height: var(--radix-popper-available-height); --radix-popover-trigger-width: var(--radix-popper-anchor-width); --radix-popover-trigger-height: var(--radix-popper-anchor-height); pointer-events: auto;"
  tabindex="-1"
>
  <div
    class="bg-white dark:bg-neutral-700"
  >
    <div
      class="relative w-full mb-1"
    >
      <span
        class="icon-solid-magnifying-glass shrink-0 absolute left-3 top-1/2 block -translate-y-1/2 text-base leading-none text-neutral-400 dark:text-white"
        role="img"
      />
      <input
        class="w-full rounded border border-neutral-250 pl-10 pr-6 text-neutral-400 placeholder:text-neutral-350 focus:border-brand-400 focus:outline-none focus:transition-[border-color] dark:border-neutral-350 dark:bg-transparent dark:text-white dark:placeholder:text-neutral-250 h-9 text-xs"
        data-testid="input-search"
        name="search"
        placeholder="Search..."
        type="text"
      />
    </div>
  </div>
  <div
    class="overflow-y-auto"
  >
    <div
      class="flex text-sm font-medium rounded-sm outline-none dark:text-neutral-100 data-[highlighted]:bg-brand-50 data-[highlighted]:text-brand-500 dark:data-[highlighted]:bg-neutral-550 cursor-pointer text-neutral-400 hover:bg-brand-50 hover:text-brand-500 h-[52px] flex-col items-start justify-center gap-1 px-2 py-1.5"
      data-orientation="vertical"
      data-radix-collection-item=""
      role="menuitem"
      tabindex="-1"
      type="button"
    >
      <span
        class="text-sm font-medium"
      >
        <span
          class=""
          data-state="closed"
          data-testid="truncate-text"
        >
          QOVERY_CONTAINER_Z04308D
          ...
        </span>
      </span>
      <span
        class="truncate text-xs font-normal"
      >
        test
      </span>
    </div>
  </div>
</div>
`;
